examples/sir_hmc.py

```diff
@@ -0,0 +1,457 @@
+    prob_s = pyro.sample("prob_s", dist.Uniform(0, 1))
+    prob_i = pyro.sample("prob_i", dist.Uniform(0, 1))
+    rho = pyro.sample("rho", dist.Uniform(0, 1))
+
+    # Sequentially filter.
+    S_curr = torch.tensor(population - 1.)
+    I_curr = torch.tensor(1.)
+    for t, datum in enumerate(data):
+        # Sample reparameterizing variables.
+        # Note the density is ignored; distributions are used only for initialization.
+        with poutine.mask(mask=False):
+            S_prev, I_prev = S_curr, I_curr
+            S_curr = pyro.sample("S_{}".format(t), dist.Binomial(population, 0.5))
+            I_curr = pyro.sample("I_{}".format(t), dist.Binomial(population, 0.5))
+
+        # Now we reverse the computation.
+        S2I = S_prev - S_curr
+        I2R = I_prev - I_curr + S2I
+        pyro.sample("S2I_{}".format(t),
+                    dist.ExtendedBinomial(S_prev, prob_s * I_prev / population),
+                    obs=S2I)
+        pyro.sample("I2R_{}".format(t),
+                    dist.ExtendedBinomial(I_prev, prob_i),
+                    obs=I2R)
+        pyro.sample("obs_{}".format(t),
+                    dist.ExtendedBinomial(S2I.clamp(min=0), rho),
+                    obs=datum)
+
+
+# By reparameterizing, we have converted to coordinates to make the model
+# Markov. We have also replaced dynamic integer_interval constraints with
+# easier static integer_interval constraints (although we'll still need good
+# initialization to avoid NANs). Since the discrete latent variables are
+# bounded (by population size), we can enumerate out discrete latent variables
+# and perform HMC inference over the global latents. However enumeration
+# complexity is O(population^4), so this is only feasible for very small
+# populations.
+#
+# Here is an inference approach using an MCMC sampler.
+
+def infer_hmc_enum(args, data):
+    model = reparameterized_discrete_model
+    return _infer_hmc(args, data, model)
+
+
+def _infer_hmc(args, data, model, init_values={}):
+    kernel = NUTS(model,
+                  max_tree_depth=args.max_tree_depth,
+                  init_strategy=init_to_value(values=init_values),
+                  jit_compile=args.jit, ignore_jit_warnings=True)
+    energies = []
+
+    def hook_fn(kernel, *unused):
+        e = float(kernel._potential_energy_last)
+        energies.append(e)
+        if args.verbose:
+            print("potential = {:0.6g}".format(e))
+
+    mcmc = MCMC(kernel, hook_fn=hook_fn,
+                num_samples=args.num_samples,
+                warmup_steps=args.warmup_steps)
+    mcmc.run(data, population=args.population)
+    mcmc.summary()
+    if args.plot:
+        import matplotlib.pyplot as plt
+        plt.figure()
+        plt.plot(energies)
+        plt.xlabel("MCMC step")
+        plt.ylabel("potential energy")
+        plt.tight_layout()
+
+    samples = mcmc.get_samples()
+    return samples
+
+
+# To perform exact inference on large populations, we'll continue to
```

Comment:
nit: asymptotically exact? i think you want to say that the reparameterization is lossless/doesn't introduce bias

Reply:
I'm catering to you here Martin :smile:
Fritz: I'll use SVI.
Martin: I doubt approximate inference will work. Try HMC instead.